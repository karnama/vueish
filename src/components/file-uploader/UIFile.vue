<template>
    <div class="flex items-center justify-between dark:text-white">
        <div class="flex items-center flex-row overflow-hidden">
            <UISpinnerLoader :determinate="!isLoading" :diameter="50" :stroke="2">
                <UIAvatar :src="src" class="avatar text-xs" :content="extension" />
            </UISpinnerLoader>
            <p :title="file.name" class="ml-4 truncate">
                {{ file.name }}
                <br>
                <small>{{ size }}</small>
            </p>
        </div>
        <UIFadeTransition>
            <template v-if="!isLoading">
                <button v-if="!uploadOnMounted"
                        class="ml-2 p-1 transition-colors hover:text-white hover:bg-red-600 rounded"
                        @click="$emit('removed', file)"
                        v-html="clearIcon" />
                <button v-else-if="failedToUpload"
                        class="transition transform hover:rotate-180 !duration-200"
                        @click="uploadFile"
                        v-html="retryIcon" />
                <span v-else class="text-green-600 dark:text-green-500" v-html="checkIcon" />
            </template>
        </UIFadeTransition>
    </div>
</template>

<script lang="ts">
import type { PropType } from 'vue';
import { computed, defineComponent, onMounted, ref, watch } from 'vue';
import UIAvatar from '@components/avatar/UIAvatar.vue';
import { getIcon } from '@/helpers';
import UISpinnerLoader from '@components/loader-spinner/UISpinnerLoader.vue';
import UIFadeTransition from '@components/transitions/UIFadeTransition.vue';
import { getSizeString, isImage as fileIsImage, getExtension } from '@composables/utils';

// todo accept a composable (returns a ref for progress) for upload
export default defineComponent({
    name: 'UIFile',
    components: { UIAvatar, UISpinnerLoader, UIFadeTransition },
    props: {
        /**
         * The file that is being handled.
         */
        file: {
            type: File,
            required: true
        },

        /**
         * Upload as soon as it's mounted.
         */
        uploadOnMounted: {
            type: Boolean,
            default: false
        },

        /**
         * The function that uploads the files.
         */
        upload: {
            type: Function as PropType<(files: File|File[]) => Promise<void>>,
            required: true
        }
    },

    emits: ['removed'],

    setup: (props) => {
        const clearIcon = getIcon('clear');
        const checkIcon = getIcon('check');
        const retryIcon = getIcon('retry');

        const isLoading = ref(false);
        const src = ref<string | ArrayBuffer | null>();
        const failedToUpload = ref(false);

        const isImage = computed(() => fileIsImage(props.file));
        const extension = computed(() => {
            const ext = getExtension(props.file);

            return ext ? '.' + ext : 'file';
        });
        const size = computed(() => getSizeString(props.file));

        const uploadFile = async () => {
            isLoading.value = true;

            await props.upload([props.file])
                .then(resp => {
                    failedToUpload.value = false;
                    return resp;
                })
                .catch(err => {
                    failedToUpload.value = true;
                    return err;
                })
                .finally(() => isLoading.value = false);
        };

        watch(
            () => props.file,
            file => {
                if (!isImage.value) {
                    src.value = null;
                    return;
                }
                const reader = new FileReader();
                reader.readAsDataURL(file);
                reader.onloadend = () => src.value = reader.result;
            },
            { immediate: true });

        onMounted(async () => {
            if (!props.uploadOnMounted) return;

            await uploadFile();
        });

        return {
            size,
            extension,
            src,
            clearIcon,
            checkIcon,
            retryIcon,
            isLoading,
            failedToUpload,
            uploadFile
        };
    }
});
</script>

<style scoped lang="scss">
.avatar {
    width: 2.5rem;
    height: 2.5rem;
    min-width: 2.5rem;
    min-height: 2.5rem;
}
</style>
