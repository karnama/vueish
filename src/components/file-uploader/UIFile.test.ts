import { mount } from '@vue/test-utils';
import UIFile from '@components/file-uploader/UIFile.vue';
import { nextTick } from 'vue';

describe('UIFile', () => {
    const file = new File(['foo'], 'foo.txt', {
        type: 'text/plain'
    });
    const upload = () => {};

    it('should display correctly', () => {
        const wrapper = mount(UIFile, {
            props: {
                file,
                upload
            }
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it('should display the size in the kb or mb', async () => {
        const wrapper = mount(UIFile, {
            props: {
                file,
                upload
            }
        });

        expect(wrapper.html()).toContain('0.0029 KB');
        const newFile = new File([Array(100000).fill('foo').join()], 'foo.txt', {
            type: 'text/plain'
        });

        await wrapper.setProps({ file: newFile });
        expect(wrapper.html()).toContain('0.38 MB');
    });

    it('should emit an event with the removed file on remove', async () => {
        const wrapper = mount(UIFile, {
            props: {
                file,
                upload
            }
        });

        await wrapper.get('button').trigger('click');

        expect(wrapper.lastEventValue<[File]>('removed')![0].size).toStrictEqual(file.size);
    });

    // tried mocking the reader.result too to no avail
    // eslint-disable-next-line jest/no-disabled-tests
    it.skip('should display an image if the file is an image', async () => {
        const base64 = 'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABAQMAAAAl21bKAAAAA1BMVEUAAACnej3a' +
            'AAAAAXRSTlMAQObYZgAAAApJREFUCNdjYAAAAAIAAeIhvDMAAAAASUVORK5CYII=';
        const imgFile = new File(['data:image/png;base64,' + base64], '1x1.png', {
            type: 'image/png'
        });
        const wrapper = mount(UIFile, {
            props: {
                file: imgFile,
                upload
            }
        });

        expect(wrapper.find('img').isVisible()).toBe(true);
        await wrapper.setProps({ file });
        expect(wrapper.find('img').isVisible()).toBe(false);
    });

    it('should call the given upload if uploading on mounted', () => {
        const upload = jest.fn();
        mount(UIFile, {
            props: {
                file,
                upload
            }
        });

        expect(upload).not.toHaveBeenCalled();

        mount(UIFile, {
            props: {
                file,
                upload,
                uploadOnMounted: true
            }
        });

        expect(upload).toHaveBeenCalledWith(file);
    });
});
