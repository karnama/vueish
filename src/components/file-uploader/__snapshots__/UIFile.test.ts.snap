// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIFile should display correctly 1`] = `
<div
  class="flex items-center justify-between dark:text-white"
>
  <div
    class="flex items-center flex-row overflow-hidden"
  >
    <transition-stub>
      <div
        class="ui-spinner-loader text-brand-400"
        role="progressbar"
      >
        <svg
          class="spinner-draw stroke-current"
          focusable="false"
          preserveAspectRatio="xMidYMid"
          style="width: 50px; height: 50px;"
          viewBox="0 0 50 50"
        >
          <circle
            class="spinner-circle stroke-current"
            cx="50%"
            cy="50%"
            r="24"
            stroke-width="2px"
            style="stroke-dashoffset: 150.79644737231007px; stroke-dasharray: 150.79644737231007px; --spinner-start-value: 143.25662500369455; --spinner-end-value: 30.159289474462014;"
          />
        </svg>
        <div
          class="abs-center"
        >
          
          <div
            class="ui-avatar text-lg overflow-hidden min-h-[2rem] min-w-[2rem] h-8 w-8 text-white bg-brand-400 flex flex-col justify-center items-center rounded-full avatar text-xs"
          >
            
            
            .txt
            
            
          </div>
          
        </div>
      </div>
    </transition-stub>
    <p
      class="ml-4 truncate"
      title="foo.txt"
    >
      foo.txt 
      <br />
      <small>
        3B
      </small>
    </p>
  </div>
  <transition-stub>
    
    
    <button
      class="ml-2 p-1 transition-colors hover:text-white hover:bg-red-600 rounded"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 20 20"
        width="20px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    
    
  </transition-stub>
</div>
`;
